### PR TITLE
Remove managed interfaces impls defined in CoreCLR

### PR DESCRIPTION
**Pull request by AaronRobinsonMSFT**
The `IReflect` implementation can be removed since we missed a supporting type after we brought this back in https://github.com/dotnet/coreclr/pull/21343. The supporting type is `System.Runtime.InteropServices.ComTypes.IReflect`, which is needed for it to work at all. Thus, this scenario has never worked in .NET Core.

Implement `IDynamicInterfaceCastable` on `__ComObject` to avoid managed interface definitions in unmanaged code - `IEnumerable` and `IEnumerator`. In an effort to ensure existing users are supported, `ICustomAdapter` is available when either of the previous two are supported. This allows us to remove the FCall forwarders that simulated a type of DIM for RCWs that implemented a conforming `IDispatch` or `IEnumVARIANT`.

Verified that `EnumerableToDispatchMarshaler` is not emitted by TlbImp.

### CODE REVIEW

**Review by jkotas — APPROVED** [review]
Nice!

File: src/coreclr/vm/mngstdinterfaces.h [hunk 1]
```diff
@@ -1,397 +0,0 @@
...
-            return NULL;
-        }
-    }
-
-private:
-    // Disalow creation of this class by anybody outside of it.
-    MngStdInterfaceMap();
-
-    // The map of type names to native IID's.
-    EEUtf8StringHashTable m_TypeNameToNativeIIDMap;
```
[the remark this answers is not in the document]
Comment (AaronRobinsonMSFT) [reply src/coreclr/vm/mngstdinterfaces.h:85]: Sure.

File: src/coreclr/System.Private.CoreLib/src/System/__ComObject.cs [hunk 2]
```diff
@@ -135,11 +141,185 @@ private object CreateEventProvider(
...
+                return IsIEnumerable(this);
+            }
+            else if (interfaceType.Equals(typeof(IEnumerator).TypeHandle))
+            {
+                return IsIEnumerator(this);
+            }
+            else if (interfaceType.Equals(typeof(ICustomAdapter).TypeHandle))
+            {
+                // This is for backwards compatibility purposes.
+                return IsIEnumerator(this) || IsIEnumerable(this);
```
[the remark this answers is not in the document]
Comment (AaronRobinsonMSFT) [reply src/coreclr/System.Private.CoreLib/src/System/__ComObject.cs:164]: Done.

File: src/coreclr/System.Private.CoreLib/src/System/__ComObject.cs [hunk 3]
```diff
@@ -135,11 +141,185 @@ private object CreateEventProvider(
...
+            }
+            else
+            {
+                return false;
+            }
+
+            static bool IsIEnumerable(__ComObject co)
```
[the remark this answers is not in the document]
Comment (AaronRobinsonMSFT) [reply src/coreclr/System.Private.CoreLib/src/System/__ComObject.cs:171]: I think that is already done in 

https://github.com/dotnet/runtime/blob/f826ef2b41779374a2a820154f9414ce4ff215a1/src/tests/Interop/COM/NETClients/IDispatch/Program.cs#L193-L195

File: src/tests/Interop/PInvoke/IEnumerator/IEnumeratorTest.cs [hunk 4]
```diff
@@ -61,6 +61,7 @@ public static void TestManagedToNative()
         public static void TestNativeRoundTrip()
         {
             IEnumerator nativeEnumerator = IEnumeratorNative.GetIntegerEnumerator(1, 10);
+            Assert.True(nativeEnumerator is ICustomAdapter);
```
Comment (AaronRobinsonMSFT) [inline src/tests/Interop/PInvoke/IEnumerator/IEnumeratorTest.cs:64]:
```suggestion
```

File: src/coreclr/System.Private.CoreLib/src/System/__ComObject.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (AaronRobinsonMSFT) [reply src/coreclr/System.Private.CoreLib/src/System/__ComObject.cs:171]: Ah, I see. Sure, let me add a case for that.

**Review by elinor-fung — APPROVED** [review]
Yay!

**Comment by AaronRobinsonMSFT** [comment]
Breaking change issue filed at https://github.com/dotnet/docs/issues/49921

**Pull request merged**